File: Multi_Threading.py
Python threading tutorial
 ******************************************************
 thread =  a flow of execution. Like a separate order of instructions.
                  However each thread takes a turn running to achieve concurrency
                  GIL = (global interpreter lock),
                  allows only one thread to hold the control of the Python interpreter at any one time
      
      THERE ARE TWO TYPES OF TASKS
      
 io bound = program/task spends most of its time waiting for external events (user input, web scraping)
            use multithreading


 cpu bound = program/task spends most of its time waiting for internal events (CPU intensive)
             use multiprocessing


  
  HERE WE ARE GOING TO TALK ABOUT MULTI THREADING
  
import threading
import time

***********************************************
Here we create three functions that dleay for some seconds

def eat_breakfast():
    time.sleep(3)
    print("You eat breakfast")


def drink_coffee():
    time.sleep(4)
    print("You drank coffee")


def study():
    time.sleep(5)
    print("You finish studying")
    
    
***********************************************
    Here we start threads to run the differnet functions. Then we use  the start function to start that thread
    


x = threading.Thread(target=eat_breakfast, args=())   // This creates a thread and runs that function.We use the args to pas any arguments that that function uses
x.start()                                            // This starts that thread

y = threading.Thread(target=drink_coffee, args=())
y.start()

z = threading.Thread(target=study, args=())
z.start()
  




***********************************************


print(threading.active_count())   // We use this to tell us the number of active and running threads
print(threading.enumerate())    // We use this to ive us the name of the threads that are running
print(time.perf_counter())     // We use this to count the time a thread takes to excute code


NOTE----> ALL CODE RUNS ON THE MAIN THREAD , UNLESS THE THREAD IS CHANGED LIKE ABOVE.
tHERE IS ALSO SOMETHING CALLED THREAD SYNCHLONISATION

This is where we make or main thread wait for other threads to finish excuting their code

To do this 


x.join()  // This makes the the main thread to wait for the x thread to finish its code excutinn
y.join() ///The main thread also waits for the y 
z.join()  /// and x threads to finsh code excution






HERE IS MORE EXPLANATION
------------------------------------------------------------------------------------------------------------------------------

Multithreading in Python allows you to run multiple threads (lightweight processes) concurrently within a single process. 
This can improve the performance of your program by allowing it to execute multiple tasks simultaneously,
especially when some of the tasks involve I/O operations that would otherwise cause the program to block.

Python provides several built-in modules for working with multithreading, including threading,
queue, and concurrent.futures. The threading module is the most commonly used module for working with threads in Python.

Heres an example that demonstrates how to create and run a simple multithreaded program using the threading module:


import threading
import time

def worker():
    print(f"Worker {threading.get_ident()} started")
    time.sleep(2)
    print(f"Worker {threading.get_ident()} finished")

# Create a list of worker threads
threads = []
for i in range(5):
    t = threading.Thread(target=worker)
    threads.append(t)

# Start the threads
for t in threads:
    t.start()

# Wait for the threads to finish
for t in threads:
    t.join()

print("All workers finished")




In this example, we define a worker() function that simulates some work by sleeping for 2 seconds.
We then create a list of 5 worker threads using the Thread class from the threading module, and we start the threads using the start() method. 
Finally, we wait for all of the threads to finish using the join() method, and we print a message indicating that all of the workers have finished.

When you run this program, you should see output similar to the following:

Worker 140285925769728 started
Worker 140285933162432 started
Worker 140285916376320 started
Worker 140285924986624 started
Worker 140285933665536 started
Worker 140285925769728 finished
Worker 140285933162432 finished
Worker 140285916376320 finished
Worker 140285924986624 finished
Worker 140285933665536 finished
All workers finished



Note that the threading.get_ident() function is used to get the unique identifier for each thread,
which allows us to distinguish between the different threads in the output.




























































































..

File: Multi_threading.kt
            thread = an execution path of a program
                     We can use multiple threads to perform,different tasks of our program at the same time.
                     Current thread running is "main" thread
                     using System.Threading;  --> The namespace that enables us to create and run threads




using System;
using System.Threading;

namespace MyFirstProgram
{
    class Program
    {
        static void Main(string[] args)
        {
            // thread = an execution path of a program
            //          We can use multiple threads to perform,
            //          different tasks of our program at the same time.
            //          Current thread running is "main" thread
            //          using System.Threading;

            Thread mainThread = Thread.CurrentThread;
            mainThread.Name = "Main Thread";
            //Console.WriteLine(mainThread.Name);

            Thread thread1 = new Thread(() => CountDown("Timer #1"));   / We use lamda incase there is an argument we have to pass to the function
            Thread thread2 = new Thread(CountUP);                      /  If ther are no arguments, then we just put the name of the function , instead of using a lambda function           

            thread1.Start();
            thread2.Start();

            Console.WriteLine(mainThread.Name + " is complete!");

            Console.ReadKey();
        }  
        public static void CountDown(String name)
        {
            for (int i = 10; i >= 0; i--)
            {
                Console.WriteLine("Timer #1 : " + i + " seconds");
                Thread.Sleep(1000);
            }
            Console.WriteLine("Timer #1 is complete!");
        }
        public static void CountUp()
        {
            for (int i = 0; i <= 10; i++)
            {
                Console.WriteLine("Timer #2 : " + i + " seconds");
                Thread.Sleep(1000);
            }
            Console.WriteLine("Timer #2 is complete!");
        }
    }
}








MORE EXPLANATION
*********************************************************************************************************************************************************************

Multithreading in C# is a process of executing multiple threads simultaneously, which allows programs to perform multiple tasks concurrently.
Threads are lightweight sub-processes and are the smallest unit of processing.
Using multithreading can improve the performance of an application and make it more responsive javatpoint.com.

To create and manage threads in C#, you can use the System.Threading namespace. 
The Thread class in this namespace provides methods to create and manipulate threads


Multithreading is a technique that allows a single program to run multiple tasks simultaneously. In C#, threads are created using the Thread class. To create a thread, you can use the following code:

Code snippet
Thread myThread = new Thread(myThreadFunction);
myThread.Start();
Use code with caution. Learn more
The myThreadFunction parameter is a delegate that represents the task that you want the thread to perform. 
The Start() method starts the thread running.

Once a thread has been started, it will continue to run until it completes its task or is interrupted.
You can interrupt a thread using the Interrupt() method.

To use a thread, you can use the Thread.CurrentThread property to get a reference to the current thread.
You can then use this reference to call methods on the thread, such as Sleep() and Join().

The Sleep() method pauses the thread for a specified amount of time. The Join() method waits for the thread to complete before continuing execution.

Here is an example of a simple multithreaded program:

Code snippet
using System;
using System.Threading;

namespace MultithreadingDemo
{
    class Program
    {
        static void Main(string[] args)
        {
            // Create a new thread.
            Thread myThread = new Thread(myThreadFunction);

            // Start the thread running.
            myThread.Start();

            // Wait for the thread to complete.
            myThread.Join();

            // The thread has now completed, so we can print a message.
            Console.WriteLine("The thread has completed.");
        }

        static void myThreadFunction()
        {
            // Do some work in the thread.
            for (int i = 0; i < 10; i++)
            {
                Console.WriteLine("Thread is running...");
                Thread.Sleep(1000);
            }
        }
    }
}
Use code with caution. Learn more
This program creates a new thread and starts it running. The thread then loops 10 times and prints 
a message to the console. Once the thread has completed, the main thread prints a message to the console.

Multithreading can be used to improve the performance of your programs by allowing them to perform multiple tasks simultaneously. 
However, it is important to use multithreading carefully, as it can also introduce problems such as race conditions and deadlocks.

Here are some of the uses and applications of multithreading in C#:

Improving the performance of programs by allowing them to perform multiple tasks simultaneously.
Creating user interfaces that are responsive and interactive.
Downloading files in the background while the user continues to work in the foreground.
Compressing or decompressing files in the background.
Encrypting or decrypting data in the background.
Parsing or searching large amounts of data.
Simulating real-world events, such as traffic or weather.
Running multiple games or applications simultaneously.
Multithreading can be a powerful tool for improving the performance and functionality of your C# programs.
However, it is important to use it carefully to avoid problems such as race conditions and deadlocks.




WHAT IS A DEAD LOCK
-------------


A deadlock is a situation in which two or more threads are blocked indefinitely, 
waiting for each other to release a resource. This can happen when two threads are both trying to access the same resource, 
and each thread is holding a resource that the other thread needs.

In the above context, a deadlock could occur if two threads were both trying to access the same file. 
If one thread had the file open for reading, and the other thread had the file open for writing, then the two threads would be deadlocked. 
Neither thread would be able to continue until the other thread released the file.

Deadlocks can be difficult to diagnose and resolve. One way to prevent deadlocks is to avoid having threads share resources. 
If threads must share resources, then it is important to use locks to ensure that only one thread can access a resource at a time.

Here are some of the ways to avoid deadlocks:

Avoid having threads share resources.
Use locks to ensure that only one thread can access a resource at a time.
Use a deadlock detection and prevention algorithm.
Use a deadlock avoidance algorithm.

Deadlocks can be a serious problem, as they can cause programs to hang or crash. 
It is important to be aware of the potential for deadlocks and to take steps to prevent them.









MORE EXPLANATION
*********************************************************************************************************************************************************************


  Multithreading in C# refers to the ability to execute multiple threads concurrently within a single application.
A thread represents an independent path of execution within a program, allowing different parts of the code to execute simultaneously. 
Multithreading is used to improve performance, responsiveness, and resource utilization in applications that can benefit from parallel processing.

To create and use threads in C#, you can utilize the System.Threading namespace. 
Heres an example of creating and starting a new thread:


using System;
using System.Threading;

class Program
{
    static void Main()
    {
        // Create a new thread and specify the method to be executed
        Thread thread = new Thread(WorkerMethod);

        // Start the thread
        thread.Start();

        // Main thread continues execution
        Console.WriteLine("Main thread executing...");

        // Wait for the worker thread to complete
        thread.Join();

        Console.WriteLine("Worker thread completed. Main thread exiting.");
    }

    static void WorkerMethod()
    {
        // Code executed by the worker thread
        Console.WriteLine("Worker thread executing...");
        Thread.Sleep(2000); // Simulate some work
        Console.WriteLine("Worker thread completed.");
    }
}
In the above code, we create a new thread using the Thread class and specify the method WorkerMethod that will be executed by the thread. 
We start the thread using the Start() method, and the main thread continues its execution. 
The worker thread executes its own code in parallel and then completes. 
Finally, the main thread waits for the worker thread to finish using the Join() method.

Multithreading has various uses and applications in C#:

Concurrent Execution: Multithreading allows you to execute different parts of your code simultaneously,
enabling concurrent execution of tasks and improving overall performance and responsiveness.

Asynchronous Operations: Multithreading is used for performing time-consuming or I/O-bound operations asynchronously,
ensuring that the application remains responsive to user interactions.

Parallel Processing: Multithreading facilitates parallel processing of computationally intensive tasks,
allowing you to split the workload among multiple threads and leverage multiple CPU cores for improved performance.

Background Tasks: Multithreading is employed for running background tasks, such as data processing, file downloads, 
or network communication, while the main thread handles the user interface and remains responsive.

Resource Utilization: Multithreading helps maximize resource utilization, such as utilizing multiple CPU cores efficiently 
and making better use of available system resources.

Its important to note that multithreading introduces challenges related to thread synchronization,
shared data access, and potential race conditions. Proper synchronization mechanisms, such as locks, monitors, 
and thread-safe collections, should be used to ensure thread safety and avoid data corruption.

Multithreading in C# provides a powerful mechanism to improve application performance, responsiveness, 
And resource utilization. It allows you to leverage concurrent execution, asynchronous operations, 
and parallel processing for better scalability and enhanced user experience. However, it also requires careful consideration 
of thread safety and synchronization to avoid potential issues.












































































































































































...
